src/Control/Monad/Aff/Bus.purs

```diff
@@ -50,20 +55,38 @@ type BusRW = Bus (read ∷ Cap, write ∷ Cap)
 
 -- | Creates a new bidirectional Bus which can be read from and written to.
 make ∷ ∀ eff a. Aff (avar ∷ AVAR | eff) (BusRW a)
-make = Bus <$> makeEmptyVar
+make = do
+  cell ← makeEmptyVar
+  consumers ← makeVar mempty
+  let
+    loop = do
+      attempt (takeVar cell) >>= traverse_ \res → do
+        vars ← takeVar consumers
+        putVar mempty consumers
+        sequence_ (foldl (\xs a → putVar res a : xs) mempty vars)
+        loop
+  _ ← forkAff loop
+  pure $ Bus cell consumers
 
 -- | Blocks until a new value is pushed to the Bus, returning the value.
 read ∷ ∀ eff a r. BusR' r a → Aff (avar ∷ AVAR | eff) a
-read (Bus avar) = readVar avar
+read (Bus _ consumers) = do
+  res' ← makeEmptyVar
+  cs ← takeVar consumers
+  putVar (res' : cs) consumers
+  takeVar res'
 
 -- | Pushes a new value to the Bus, yieldig immediately.
 write ∷ ∀ eff a r. a → BusW' r a → Aff (avar ∷ AVAR | eff) Unit
-write a (Bus avar) = tryPutVar a avar *> void (takeVar avar)
+write a (Bus cell _) = putVar a cell
 
 -- | Splits a bidirectional Bus into separate read and write Buses.
 split ∷ ∀ a. BusRW a → Tuple (BusR a) (BusW a)
-split (Bus avar) = Tuple (Bus avar) (Bus avar)
+split (Bus a b) = Tuple (Bus a b) (Bus a b)
 
 -- | Kills the Bus and propagates the exception to all consumers.
-kill ∷ ∀ eff a r. Error → BusW' r a → Aff (avar ∷ AVAR | eff) Unit
-kill err (Bus avar) = killVar err avar
+kill ∷ ∀ eff a r. Exn.Error → BusW' r a → Aff (avar ∷ AVAR | eff) Unit
+kill err (Bus cell consumers) = do
+  killVar err cell
+  vars ← takeVar consumers
+  traverse_ (killVar err) vars
```

Comment:
This used to kill `consumers` as well, but doing so propagated an error all the way up now, so we dropped that part. It'll be empty anyway, so I think it's still fine.